### PR TITLE
Fix/rewrite ZebraOSS SDK 

### PR DESCRIPTION
**Pull request by DaSpood**
- Reorganize packages
- Rewrite SSI parser and packet handling
- Rewrite Multi-packet support (supports DECODE_DATA only)
- Update handling of DECODE_DATA (scans) to fit new findings in the SSI protocol
- Add support for BLE with SSI-over-ATT (requires more testing)

### CODE REVIEW

File: enioka_scan/src/main/java/com/enioka/scanner/bt/manager/ClassicBtSppScanner.java [hunk 1]
```diff
@@ -336,10 +336,15 @@ public void registerStatusCallback(ScannerStatusCallback statusCallback) {
...
+    @Override
+    public boolean isBleDevice() {
+        return false;
+    }
+
     void handleInputBuffer(byte[] buffer, int offset, int length) {
         int read = offset;
         while (read < length) {
-            read = handleInputBufferLoop(buffer, read, length) + 1;
+            read += handleInputBufferLoop(buffer, read, length); // FIXME: doubtful about the +1 ? If X bytes are read, read=X, so next start offset will be X which should be correct if we start at 0
```
Comment (marcanpilami) [inline enioka_scan/src/main/java/com/enioka/scanner/bt/manager/ClassicBtSppScanner.java:347]: Ca ressemble en effet à un gros bug

File: enioka_scan/src/main/java/com/enioka/scanner/sdk/zebraoss/SsiOverSppParser.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+            packetLength = (char) (buffer[offset] & 0xFF);
+            packetLength += 2; // include checksum
+            ssiBuffer = new byte[packetLength];
+        }
+
+        System.arraycopy(buffer, offset, ssiBuffer, read, Math.min(dataLength - offset, packetLength - read));
+        read += Math.min(dataLength, packetLength - read);
+        res.read = read;
+
+        if (read == packetLength) { // SSI packet is complete
```
Comment (marcanpilami) [inline enioka_scan/src/main/java/com/enioka/scanner/sdk/zebraoss/SsiOverSppParser.java:31]: pourquoi re-bufferiser ici ? PLutôt que retourner directement le résultat de SsiParser ?

File: enioka_scan/src/main/java/com/enioka/scanner/sdk/zebraoss/ssi/SsiMonoPacket.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+/**
+ * SSI packet, containing the header and checksum bytes. Suitable for both incoming and outgoing packets.
+ * Multi-packets are (so far) read-only and in a separate class as they require extra header elements.
+ */
+public final class SsiMonoPacket {
+    private static final String LOG_TAG = "SsiMonoPacket";
+    private final boolean ble;
+
+    // Packet attributes are in the same order as which they occur in the raw packet buffer
+    /** MSB of total length in little-endian used by BLE only, for outgoing packets only (these bytes should be skipped for reading). */
```
Comment (marcanpilami) [inline enioka_scan/src/main/java/com/enioka/scanner/sdk/zebraoss/ssi/SsiMonoPacket.java:12]: why skipped? Aren't they useful to detect if reading is complete?

File: enioka_scan/src/main/java/com/enioka/scanner/sdk/zebraoss/ssi/SsiMonoPacket.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
+package com.enioka.scanner.sdk.zebraoss.ssi;
+
+/**
+ * SSI packet, containing the header and checksum bytes. Suitable for both incoming and outgoing packets.
+ * Multi-packets are (so far) read-only and in a separate class as they require extra header elements.
+ */
+public final class SsiMonoPacket {
+    private static final String LOG_TAG = "SsiMonoPacket";
```
Comment (marcanpilami) [inline enioka_scan/src/main/java/com/enioka/scanner/sdk/zebraoss/ssi/SsiMonoPacket.java:8]: we tend only to have a single log tag per sdk to make debugging easier

File: enioka_scan/src/main/java/com/enioka/scanner/sdk/zebraoss/ssi/SsiMultiPacket.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+    /**
+     * Reads a packet that is part of a multi-packet batch.
+     * @param buffer The buffer containing the raw packet data, including headers and checksum
+     * @param offset The offset from which packet data starts. In case of BLE, the offset must skip the first two bytes of the packet.
+     * @param length The length of the packet data, including headers and checksum
+     */
+    public void readPacket(final byte[] buffer, int offset, int length) {
+        if (!expectingMoreData)
+            throw new IllegalStateException("Not expecting more data");
```
Comment (marcanpilami) [inline enioka_scan/src/main/java/com/enioka/scanner/sdk/zebraoss/ssi/SsiMultiPacket.java:42]: style: always use braces (everywhere, not only here ;-))

File: enioka_scan/src/main/java/com/enioka/scanner/sdk/zebraoss/ssi/SsiMultiPacket.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+import java.nio.ByteOrder;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * SSI multi-packet, containing the header and checksum bytes. Suitable for both incoming.
+ * So far, only DECODE_DATA packets seem to sometimes use multi-packets for large barcodes, so this
+ * class is only used as a buffer for reading those types of packets before being converted to a
+ * barcode with no other use. For this reason, many sanity checks are skipped.
+ */
+public final class SsiMultiPacket {
```
Comment (marcanpilami) [inline enioka_scan/src/main/java/com/enioka/scanner/sdk/zebraoss/ssi/SsiMultiPacket.java:17]: no test it seems for multi packet, please add one from your captures

File: enioka_scan/src/main/java/com/enioka/scanner/sdk/zebraoss/ssi/SsiMonoPacket.java [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+        //this.validateLengthAndChecksum(); // FIXME - 2022/03/28: inbound packets are constantly rejected due to checksum. For now, assume they are all correct.
+    }
+
+    /**
+     * Creates a new SSI packet for sending commands.
+     * Only suitable for outgoing packets. Checksum and length are calculated automatically.
+     * @param ble Whether the packet must contain BLE length bytes or not.
+     */
+    public SsiMonoPacket(final byte opCode, final byte status, final byte[] data, final boolean ble) {
```
Comment (marcanpilami) [inline enioka_scan/src/main/java/com/enioka/scanner/sdk/zebraoss/ssi/SsiMonoPacket.java:62]: I have a big issue with this `ble` new argument, that is only here to knwo if packets should be cut into smaller pieces with a header. This means that every command must know this, and therefore that any user of the command must know if BLE or not. This is not a concern for the caller really. We should find a way to pass that infomration without modifying the API of every command.

**Review by marcanpilami — CHANGES_REQUESTED** [review]

File: enioka_scan/src/main/java/com/enioka/scanner/sdk/zebraoss/SsiOverSppParser.java (hunk 2 above)
Comment (DaSpood) [reply enioka_scan/src/main/java/com/enioka/scanner/sdk/zebraoss/SsiOverSppParser.java:31]: SPP peut être "décallé" dans les deux sens, dans certains cas un packet SPP contient plus d'un packet SSI (auquel cas on pourrait effectivement passer directement par SsiParser), dans d'autres cas il contient un packet SSI incomplet, auquel cas il peut manquer des informations au parser. Bufferiser ici était une manière plus élégante de garantir que le SsiParser ne reçoive que des paquets SSI complets, les SsiOverSpp/AttParser se chargeant d'aggréger les packets SPP/ATT

File: enioka_scan/src/main/java/com/enioka/scanner/sdk/zebraoss/ssi/SsiMultiPacket.java (hunk 6 above)
Comment (DaSpood) [reply enioka_scan/src/main/java/com/enioka/scanner/sdk/zebraoss/ssi/SsiMultiPacket.java:17]: done

File: enioka_scan/src/main/java/com/enioka/scanner/sdk/zebraoss/ssi/SsiMonoPacket.java (hunk 3 above)
Comment (DaSpood) [reply enioka_scan/src/main/java/com/enioka/scanner/sdk/zebraoss/ssi/SsiMonoPacket.java]: No longer used

**Review by marcanpilami — APPROVED** [review]

**Pull request merged**